File: 10-top-k-frequent/solution.cpp
// Top K Frequent Elements
// Time: O(n log k), Space: O(n) using min heap

#include <vector>
#include <unordered_map>
#include <queue>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // Count frequencies
        unordered_map<int, int> freqMap;
        for (int num : nums) {
            freqMap[num]++;
        }
        
        // Use min heap of size k
        auto cmp = [](pair<int, int>& a, pair<int, int>& b) {
            return a.second > b.second;
        };
        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(cmp)> minHeap(cmp);
        
        for (auto& entry : freqMap) {
            minHeap.push(entry);
            if (minHeap.size() > k) {
                minHeap.pop();
            }
        }
        
        // Extract results
        vector<int> result;
        while (!minHeap.empty()) {
            result.push_back(minHeap.top().first);
            minHeap.pop();
        }
        
        return result;
    }
};

// Test
int main() {
    Solution sol;
    vector<int> nums = {1, 1, 1, 2, 2, 3};
    vector<int> result = sol.topKFrequent(nums, 2);
    cout << "[";
    for (int i = 0; i < result.size(); i++) {
        cout << result[i];
        if (i < result.size() - 1) cout << ", ";
    }
    cout << "]" << endl; // Output: [1, 2] or [2, 1]
    return 0;
}

File: 10-top-k-frequent/solution.go
// Top K Frequent Elements
// Time: O(n log n), Space: O(n)

package main

import (
    "container/heap"
    "fmt"
)

type Item struct {
    value int
    freq  int
}

type MinHeap []Item

func (h MinHeap) Len() int           { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i].freq < h[j].freq }
func (h MinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MinHeap) Push(x interface{}) {
    *h = append(*h, x.(Item))
}

func (h *MinHeap) Pop() interface{} {
    old := *h
    n := len(old)
    item := old[n-1]
    *h = old[0 : n-1]
    return item
}

func topKFrequent(nums []int, k int) []int {
    // Count frequencies
    freqMap := make(map[int]int)
    for _, num := range nums {
        freqMap[num]++
    }
    
    // Use min heap of size k
    h := &MinHeap{}
    heap.Init(h)
    
    for value, freq := range freqMap {
        heap.Push(h, Item{value, freq})
        if h.Len() > k {
            heap.Pop(h)
        }
    }
    
    // Extract results
    result := make([]int, k)
    for i := 0; i < k; i++ {
        result[i] = heap.Pop(h).(Item).value
    }
    
    return result
}

// Test
func main() {
    result := topKFrequent([]int{1, 1, 1, 2, 2, 3}, 2)
    fmt.Println(result) // Output: [2 1] or [1 2]
}
